coderd/prometheusmetrics/aggregator.go

```diff
@@ -129,9 +157,12 @@ func NewMetricsAggregator(logger slog.Logger, registerer prometheus.Registerer,
 		log:                    logger.Named(loggerName),
 		metricsCleanupInterval: metricsCleanupInterval,
 
+		store: map[string]annotatedMetric{},
```

Comment:
Suggestion: You could use a struct key instead of a very long string. The tags still have to be processed since a slice won't work, but I think a struct is easier to maintain/understand.

Reply:
Also, I wonder if a pointer (`*annotatedMetric`) might be more performant here? It would require testing though. Pointer will take up a little bit more memory, though, but we'll also copy less memory around. (Larger the struct, larger the benefit, I imagine).

Reply:
I will explore the structKey option, good feedback! Regarding the pointer, I used to find it dangerous as it's fairly easy to have long-lived dangling objects and cause some nasty memory leaks.

Reply:
A map of pointers is no more (and no less!) a memory leak than a map of the data directly, since Go is garbage collected.

Reply:
Ok, I switched to `metricKey`, but wasn't able to include `[]metricLabel` due to compiler complaints. Is there any smart trick to include it?

Reply:
Yeah, some form of encoding on the slice will be necessary, unfortunately.

I'll admit I don't know the exact purpose of metricLabel's, are they required to uniquely identify different metrics? It would be nice if we could avoid the label encoding/hashing unless everything else matches. My proposal is not the most beautiful, but should be performant:

```go
type metricKey struct {
	// ...
	// exclude []metricLabel
}

m := map[metricKey][string]annotatedMetric{}
if storeItem, ok := m[key]; ok {
	hash := hashLabel(...)
	if am, ok := storeItem[hash]; ok {
		// full match.
	}
}
// or
m := map[metricKey][]annotatedMetric{}
if ams, ok := m[key]; ok {
	hash := hashLabel(...)
	for _, am := range ams {
		if am.hash == hash {
			// full match.
		}
	}
}
```

Reply:
Yes, the labels and values indicate a unique metric timeseries.

In practice nearly all metrics will include labels, so I don't think we gain much performance by attempting to avoid serializing and hashing them.

Reply:
Good point @spikecurtis, the benefit would only be applied once (on new items), not worth it.

Reply:
I assume that we should leave the code as is, with labels represented with a concatenated key-value string?

Reply:
Yeah: sorted, escaped, concatenated key-value string FTW. I think we assume the labels are sorted by name on input, but good to double check this.

Reply:
Ok, I used `sort.StringSlice` to sort labels.